### ISSUE TITLE
Create API endpoint to PUT a resource instance

### ISSUE THREAD

**Issue by apeters** [archesproject/arches#3485]

**Comment by apeters** [comment]
@azaroth42 
I have a very simple working version of a PUT request for a resource in json-ld format.
If you simply GET a resource in that format and edit a value, it works.  What I’m unsure of is what to expect when a user PUTs a resource when the resource includes new tile data which won’t have a valid @id (our id’s are compomposed of a tileid and a nodeid). 
EG: “@id”: “http://localhost:8000/tile/{tile id}/node/{node id}”

An example of a tile in json-ld format:
```{
    "@id": "http://localhost:8000/tile/6efb8ac0-623c-47cb-9846-4a489c153683/node/677f303d-09cc-11e7-9aa6-6c4008b05c4c",
    "@type": "http://www.cidoc-crm.org/cidoc-crm/E41_Appellation",
    "http://www.w3.org/1999/02/22-rdf-syntax-ns#value": "TEST"
}```

How would you handle this situation?

**Comment by apeters** [comment]
@azaroth42 
Rob, is it ever the case that a graph can have more then 1 branch with the same shape?
EG:
```
            Graph
            /   \
   Edge(p1)/     \Edge(p1)
          /       \
   Node(E1)       Node(E1)
```

If so then the json-ld examples above would require node-id references to unambiguate the graph shape for use in PUT and POST requests.

EG, something like this might work:
```
{
    "@id": "http://localhost:8000/tile/6efb8ac0-623c-47cb-9846-4a489c153683/node/677f303d-09cc-11e7-9aa6-6c4008b05c4c",
    "@type": [
        "http://www.cidoc-crm.org/cidoc-crm/E41_Appellation",
        "http://localhost:8000/node/677f303d-09cc-11e7-9aa6-6c4008b05c4c"
    ],
    "http://www.cidoc-crm.org/cidoc-crm/P2_has_type": {
        "@id": "http://localhost:8000/tile/6efb8ac0-623c-47cb-9846-4a489c153683/node/677f39a8-09cc-11e7-834a-6c4008b05c4c",
        "@type": [
            "http://www.cidoc-crm.org/cidoc-crm/E55_Type",
            "http://localhost:8000/node/677f303d-09cc-11e7-9aa6-6c4008b05c4c"
        ],
        "http://www.w3.org/1999/02/22-rdf-syntax-ns#value": "ecb20ae9-a457-4011-83bf-1c936e2d6b6a"
    },
    "http://www.w3.org/1999/02/22-rdf-syntax-ns#value": "Claudio"
}
```

**Comment by azaroth42** [comment]
Yes, I can think of situations where this might occur at the _shape_ level, suck as a personal name that consists of first, middle and last name parts, where the part is distinguished by a particular RDM entry.

```
{
  "type": "Name",
  "classified_as": "aat:personal_name",
  "value": "Robert Sanderson",
  "part": [
     {
       "type": "Name",
       "value": "Robert",
       "classified_as": "aat:first_name"
    },
     {
       "type": "Name",
       "value": "Sanderson",
       "classified_as": "aat:last_name"
    },   
  ]
}
```

If the first/last name bits were branches, they would have the same shape, but different RDM collections for the `classified_as` / `P2_has_type` data type.

If two branches are 100% identical, including data types and rdm collection values and so on, is there a problem with picking one at random?  I guess there could be different cards associated with each?

As the node identifier is arches specific / transaction specific and doesn't later show up in the instance data, what about a different "magic" keyword?  `"@archesNodeId": "http://localhost:8000/node/677f303d-09cc-11e7-9aa6-6c4008b05c4c"` or similar?

And only require it in the situation where the model has ambiguity as to which branch it should generate. Meaning that if you construct your models carefully, you should be able to avoid its use entirely.

**Comment by apeters** [comment]
@azaroth42
I've thought of another scenario in which the users intentions are ambiguous.
Unlike my scenario above, in this case the graph shape has unique branches.
Here though, the user's intention is to supply data for the branch with the 3 nodes, but in this case, the user only supplied data for Nodes with type E14 and E55 (they omitted data for the E62 node with the intention of leaving it null). 
Arches in this case would assume that the user is supplying data for the 2-Node branch.  If the user doesn't supply an `"@archesNodeId"` then their intentions will be mis-interpreted and data mis-applied.  
I guess my question is, `should/can we just require that users ALWAYS supply an @archesNodeId so that we eliminate these ambiguities?`  
I'm guessing that there are other scenarios in which similar issues also arise.

```
              Graph
              /   \ 
      Node(E14)   Node(E14)
          /        /     \
   Node(E55)   Node(E55) Node(E62)
```

**Comment by azaroth42** [comment]
By requiring the nodeid, it ties the requests explicitly to Arches, making it impossible to just run a generic workflow. I agree that there can be situations like this, but we shouldn't penalize clients (and careful data modelers) that are creating unambiguous instances / models. 

Given that it would be hard to debug situations where the data goes in but is assigned to a random branch out of several options, I think the server should return an error (HTTP 400) with a message that says the request is ambiguous and to use @archesNodeId to disambiguate.

Note that the @archesNodeId trick would only work in JSON-LD ... if the data comes in as Turtle or anything else, we would need some other out of band mechanism

@workergnome -- thoughts on this thread?

**Comment by workergnome** [comment]
Following up on a conversation with @azaroth42 yesterday about this, my feeling is that implicit data, held within the structure, is really messy when we start looking at interoperability.  If there's a semantic difference between the two branches that are syntactically identical, we have a problem.  We could deal with this by treating "magic instance data" specially--which would handle the "two p2s with different values" problem, or we could require people to specify null values explicitly, which would handle the differentiating between missing and null data case.

But in all cases, we're conflating UI-level constraints with data level constraints. Having two different cards that contain the same data structure seems like an interoperability problem.

**Comment by apeters** [comment]
@workergnome I'm not sure what you're arguing for and I'm not sure what you mean when you say "magic instance data".  
Are you suggesting that users supply `"@archesNodeId"` values always in their json-ld payloads (ie: "explicit data")?

**Comment by workergnome** [comment]
I think what I was saying is that two graph structures that have identical semantics and structure, they are identical.  If we're going to say in Arches that they're actually not identical, because they're either associated with Card A or Card B, to me that means that some bit of data that is important to the entity is implicit in code (or internal data structures), and not present in the data.  

The question for me, then, is what data goes into the instance that indicates which card it belongs to?  You _could_ tie it directly to the archesID, but that seems like it's leaking implementation details into the data structurei, since it's just a pointer to the actual semantic data (which is currently managed in the internal database.)  By "Magic Instance Data", I mean what is the think that we put into the data that says "Card A", not "Card B"?

(This _doesn't_ apply to the "I've either got the branch that doesn't have this property, or I AM on the branch that has this property, but the value is NULL."  That could be handled by requiring the user to pass in an explicit NULL.)

**Comment by apeters** [comment]
@workergnome and @azaroth42
Here's another area in the same graph where json-ld input would be ambiguous without an `@archesNodeId`

Here, `Keyword` and `Record Type` are ambiguous based on @type alone.

![screen shot 2018-07-27 at 3 39 58 pm](https://user-images.githubusercontent.com/823773/43349287-a4874e24-91b3-11e8-891a-b37d451a6d78.png)

**Comment by workergnome** [comment]
Yup.  The same problem is going to exist when consuming data via the API--in the exported JSON, there's not going to be any way to distinguish between the JSON representation of those two, other than by exhaustively listing all possible Monument Types and Styles and checking if each sub-branch is typed one or the other.

We ran into this in the American Art Collaborative a bunch with LinguisticObjects--description vs. dimension statement vs. abstract.

The only solutions we came up with were lots of custom properties (worst option), lots of custom sub-classes (less terrible, but still terrible), and explicit p2_has_type values as flags.

**Comment by apeters** [comment]
Obviously there is meaning encoded in other properties of the node itself that the ontology class alone can't capture.

**Comment by apeters** [comment]
It seems that someone using this api would be doing it knowingly; that they would have to know about the semantic hierarchy of the graph itself.  If that's the case, the graph could tell them a lot (like node ids).  They could then use that when they push json-ld to Arches.

Sub-classes seems appealing, but then outside users would have to be aware of those classes and use them correctly.

**Comment by workergnome** [comment]
I think we agree--which subgraph is the correct one is based on instance property values.  Which means Arches (or the consumer) has to base their decision via introspection of the content.  

So we can solve this by embedding Arches node instance data into the graph, arches card data into the graph, or by using subclassing to encode that difference into the structure of the graph, not just the instance data. I agree there are problems with subclassing, so I'd be interested in your opinion as to which of the other is preferable.

The Linked.Art solution has been to encode these differences via the "card data" equivalent, by putting specific `p2_has_type` values in that trigger special processing.  We use AAT ids (for, say, Artwork) as those filter/trigger values, but I could see using arches-specific ontologies used there.

**Comment by azaroth42** [comment]
I think the issue is different audiences, and the extent to which they will both want to use JSON-LD to create and update instances, and the extent to which they will want to do rigorous modeling.

If you don't care about JSON-LD, then there's no problem, as you'll not run into this issue.
If you do care about JSON-LD and you do care about modeling, no problem as you'll just make models that are unambiguous.
If you **do** care about JSON-LD and you **do not** care about modeling, this is where there's the potential for the issue.

There is a hierarchy of checks:
  * determine if the class provided in the JSON-LD is the correct for the model
  * determine if the set of other properties provided are correct 
  * determine if the data types of the other properties are correct
  * determine if the RDM concepts used are correct

If all of these still do not provide a unique branch, only /then/ do we need some other disambiguating factor.  This is when I think the does care about JSON-LD and does not care so much about being rigorous with modeling comes in.  At the Getty, we are very rigorous because we can afford to be and we see the benefits of being that way. Not every Arches modeler will have that luxury.

And this is where I think an out-of-band notion of `@archesNodeId` comes in to serve this category. They can not worry so much about ambiguous edge cases by making the data specific to the internals of an Arches model. Given our rigor, I don't think we'll ever need to use it and nor should we have to. That doesn't mean that there's no use case, just that we do not have the use case.

So the final check is:
 * determine if an @archesNodeId provided is correct 

And if not, then look in configuration to either:

* Pick a branch at random, and warn about the ambiguity
* Raise an exception and halt

**Comment by apeters** [comment]
The easiest way to test this is to use Postman.
Follow these steps:
- Identify a resource type (graph) to test
- Give the anonymous user full edit privileges on the graph model
- In django admin, add the anonymous user to the "Resource Editor" group
- Identify an existing resource id to GET
- Use Postman to GET a resource in json-ld format 
   - http://localhost:8000/resources/{resource id}?format=json-ld&indent=4
- Take the output of the json-ld and copy to an editor
- Edit some of the values in the json-ld
   - if you remove any @id properties (make sure to remove all within a tile) that will be treated as a new tile
  - if you omit (remove a whole tile), then it will be deleted from the system
  - values should be updated
- Take the edited json-ld and PUT back using Postman to http://localhost:8000/resources/{resource instance id}?format=json-ld&indent=4
- You should get the updated resource back in json-ld format (make sure you get an HTTP 200)
